### PR TITLE
adding tests for #1131

### PR DESCRIPTION
**Pull request by SalehHindi**
As per #1131, I added more detailed tests for logging in and a test for the + button when adding a slice. I tried breaking the login page and the `add` function for adding a slice in `models.py` and the tests caught the breakage.

One question I had was whether to use python's assert syntax as in `assert x in y` or use unittest's assert function as in `AssertEqual(x,y)`. I see both usages across the tests so which do you prefer I use?

I will work on adding tests for adding a datasource. Are there any other tests you would like me to add? I feel like the best way to get my feet wet with a project is writing tests so I would be happy to write more tests before adding features.

Any comments appreciated and thank you for your time : )

### CODE REVIEW

**Comment by xrmx** [comment]
@SalehHindi The tests i had in mind was more to check that the ui works, e.g that there is the proper url to create a slice from a table are rendered when listing tables and such. If you can do that too it wuold be great.

**Comment by xrmx** [comment]
And to fix the tests in python3 you just have to decode the response data to a string, you can use the SupersetTestCase helper method get_resp that'll do it for you.

**Comment by SalehHindi** [comment]
@xrmx, so to be clear, you're thinking JS/Frontend tests?

**Comment by xrmx** [comment]
Nope, you are checking that url works but i think current tests already do that, what we are missing is that the rendered html contains the proper links to other pages and such.

**Comment by SalehHindi** [comment]
Gotcha, thanks! I'll add those in today. By the way, is Details link in Landscape working for you? I am trying to figure out what's wrong but I keep getting a 500 error.

**Comment by SalehHindi** [comment]
@xrmx, I made changes. I tested that on the table model view page (the page that comes after clicking the + to add a slice), that each table appears and that the show, edit, and delete buttons are there along with all their URLs.

File: tests/core_tests.py [hunk 1]
```diff
@@ -157,6 +172,39 @@ def test_slices(self):
...
+    def test_add_slice(self):
+        self.login(username='admin')
+
+        # Click on the + to add a slice
+        url = '/slicemodelview/add'
+        resp = self.get_resp(url)
+        assert 'Click on a' in resp
+        assert 'to create a Slice' in resp
+
+        tables = db.session.query(models.SqlaTable).all()
```
Comment (xrmx) [inline tests/core_tests.py]: pick the first, one should be enough

File: tests/core_tests.py [hunk 2]
```diff
@@ -157,6 +172,39 @@ def test_slices(self):
...
+            # Visit each table's explore view and make sure some
+            # important functionality is there
+            url = 'superset/explore/table/{}/'.format(table.id)
+            explore_view = self.get_resp(url)
+            assert '[{}] - untitled'.format(table.name) in explore_view
+            assert 'Datasource & Chart Type' in explore_view
+            assert 'Save as' in explore_view
+            assert 'Query' in explore_view
+
+        assert not '/explore/table/{}'.format(len(tables) + 1) in resp
```
Comment (xrmx) [inline tests/core_tests.py]: I don't get these block of tests

File: tests/core_tests.py [hunk 3]
```diff
@@ -157,6 +172,39 @@ def test_slices(self):
...
+        for table in tables:
+            # Check that the /tablemodelview/list view contains each table and
+            # their show, edit, delete buttons
+            assert table.name in resp
+            assert '/superset/explore/table/{}'.format(table.id) in resp
+            assert '/tablemodelview/show/{}'.format(table.id) in resp
+            assert '/tablemodelview/edit/{}'.format(table.id) in resp
+            assert '/tablemodelview/delete/{}'.format(table.id) in resp
+
+            # Visit each table's explore view and make sure some
```
Comment (xrmx) [inline tests/core_tests.py]: This block should be a different test, but i don't know how long this code would around given explore v2

**Review by xrmx — CHANGES_REQUESTED** [review]

File: tests/core_tests.py (hunk 1 above)
Comment (SalehHindi) [inline tests/core_tests.py]: Thanks, will do.

File: tests/core_tests.py (hunk 3 above)
Comment (SalehHindi) [inline tests/core_tests.py]: Do you have any advice for what tests I should include here? Should I not test whether each table name appears along with the edit/show/delete buttons?

File: tests/core_tests.py (hunk 2 above)
Comment (SalehHindi) [inline tests/core_tests.py]: I was testing that exactly the right number of tables appear on the page. If you imagine the case where all the correct tables appear plus extraneous tables, this test would catch that case. Do you think this is going overboard?

File: tests/core_tests.py (hunk 3 above)
Comment (xrmx) [inline tests/core_tests.py]: the tests are fine just pick the first entry as before and move them to a different test function

File: tests/core_tests.py (hunk 2 above)
Comment (xrmx) [inline tests/core_tests.py]: That's handled by flask app builder i think, no need to test that

File: tests/core_tests.py [hunk 4]
```diff
@@ -157,6 +172,26 @@ def test_slices(self):
...
+        url = '/slicemodelview/add'
+        resp = self.get_resp(url)
+        assert 'Click on a' in resp
+
+        tables = db.session.query(models.SqlaTable).all()
+        for table in tables:
+            assert table.name in resp
+            assert '/superset/explore/table/{}'.format(table.id) in resp
+
+            url = 'superset/explore/table/{}/'.format(table.id)
```
Comment (xrmx) [inline tests/core_tests.py]: drop 188-193 as the removal of explore v1 is near

File: tests/core_tests.py [hunk 5]
```diff
@@ -157,6 +172,26 @@ def test_slices(self):
...
 
+    def test_add_slice(self):
+        self.login(username='admin')
+
+        # Click on the + to add a slice
+        url = '/slicemodelview/add'
+        resp = self.get_resp(url)
+        assert 'Click on a' in resp
+
+        tables = db.session.query(models.SqlaTable).all()
```
Comment (xrmx) [inline tests/core_tests.py]: Don't need a loop, just pick the first one

File: tests/core_tests.py (hunk 4 above)
Comment (SalehHindi) [inline tests/core_tests.py]: Thanks. Done.

File: tests/core_tests.py (hunk 5 above)
Comment (SalehHindi) [inline tests/core_tests.py]: Thanks. Done.

File: superset/app/static/uploads/credentials.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+User Name,Access Key Id,Secret Access Key
```
Comment (xrmx) [inline superset/app/static/uploads/credentials.csv]: Remove this

File: superset/app/static/uploads/database.csv [hunk 7]
```diff
@@ -0,0 +1,1949 @@
+id,name,date,manner_of_death,armed,age,gender,race,city,state,signs_of_mental_illness,threat_level,flee,body_camera
```
Comment (xrmx) [inline superset/app/static/uploads/database.csv]: and this

**Review by xrmx — CHANGES_REQUESTED** [review]
Remove the qandom files you added please, the rest LGTM!

**Comment by xrmx** [comment]
To fix tests try to go straight to /tablemodelview/list/ instead of slicemodelview/add

File: superset/app/static/uploads/credentials.csv (hunk 6 above)
Comment (SalehHindi) [inline superset/app/static/uploads/credentials.csv]: Whoops, sorry about the random files, I will remove them shortly.

**Review by SalehHindi — COMMENTED** [review]
Removed random files I added by accident and got all tests and checks to pass.

**Comment by mistercrunch** [comment]
LGTM

**Pull request merged**